### PR TITLE
Extend join fuzzer to cover group execution mode

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Extend join fuzzer to cover group execution mode. To do this we split the probe and 
build input by partitioning on the join keys with one partition per each group. Then we
write the split the inputs into separate files and create table scan splits from the generated
files. For each existing query plan with table scan input, we create a corresponding
grouped execution plan.
Extend AssertQueryBuilder to support group execution configuration.

### CODE REVIEW

File: velox/exec/tests/JoinFuzzerRunner.h [hunk 1]
```diff
@@ -56,12 +58,28 @@
...
         registerVectorSerde();
     facebook::velox::filesystems::registerLocalFileSystem();
 
     facebook::velox::exec::test::joinFuzzer(seed);
     return RUN_ALL_TESTS();
   }
+
+ private:
+  // Invoked to set up memory system with arbitration.
+  static void setupMemory() {
```
Comment (duanmeng) [inline velox/exec/tests/JoinFuzzerRunner.h:72]: Could we use `ArbitratorTestUtil::createMemoryManager` here?

File: velox/exec/HashBuild.cpp [hunk 2]
```diff
@@ -500,18 +500,17 @@ void HashBuild::spillInput(const RowVectorPtr& input) {
   computeSpillPartitions(input);
 
   vector_size_t numSpillInputs = 0;
-  for (auto rowIdx = 0; rowIdx < numInput; ++rowIdx) {
-    const auto partition = spillPartitions_[rowIdx];
-    if (FOLLY_UNLIKELY(!activeRows_.isValid(rowIdx))) {
+  for (auto row = 0; row < numInput; ++row) {
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:503]: Unrelated change?

File: velox/exec/tests/JoinFuzzer.cpp [hunk 3]
```diff
@@ -164,19 +233,18 @@ bool isDone(size_t i, T startTime) {
...
       core::JoinType::kLeftSemiFilter,
       core::JoinType::kLeftSemiProject,
-      core::JoinType::kAnti,
-  };
+      core::JoinType::kAnti};
 
-  size_t idx = randInt(0, kJoinTypes.size() - 1);
-  return kJoinTypes[idx];
+  const size_t idx = randInt(0, kJoinTypes.size() - 1);
+  joinType_ = kJoinTypes[idx];
```
Comment (mbasmanova) [inline velox/exec/tests/JoinFuzzer.cpp:247]: Why is this change? It feels hard to maintain a large list of member variables with different methods changing some of these.

File: velox/exec/tests/JoinFuzzer.cpp [hunk 4]
```diff
@@ -86,55 +94,116 @@ class JoinFuzzer {
...
+      true,
+      exec::MemoryReclaimer::create())};
+  std::shared_ptr<memory::MemoryPool> writerPool_{rootPool_->addAggregateChild(
+      "joinFuzzerWriter",
+      exec::MemoryReclaimer::create())};
+
   VectorFuzzer vectorFuzzer_;
+
+  // The following fields are reset cross test iterations.
+  core::JoinType joinType_;
```
Comment (mbasmanova) [inline velox/exec/tests/JoinFuzzer.cpp:187]: This is a very long list... Will it be difficult to maintain? Is there some other way to organize the code for easier maintenance?

File: velox/exec/tests/JoinFuzzer.cpp [hunk 5]
```diff
@@ -887,6 +1074,25 @@ void JoinFuzzer::go() {
   }
 }
 
+void JoinFuzzer::cleanup() {
+  keyTypes_.clear();
```
Comment (mbasmanova) [inline velox/exec/tests/JoinFuzzer.cpp:1078]: It feels like it is easy to forget some field here.

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng The approach looks good to me. There is a lot of code changes though. It is hard to follow these.

File: velox/exec/tests/JoinFuzzer.cpp (hunk 5 above)
Comment (xiaoxmeng) [reply velox/exec/tests/JoinFuzzer.cpp:1078]: Yeah, put the test config in a separate subclass and reset all together.

File: velox/exec/tests/JoinFuzzer.cpp (hunk 4 above)
Comment (xiaoxmeng) [reply velox/exec/tests/JoinFuzzer.cpp:187]: Put it in a separate class.

File: velox/exec/HashBuild.cpp (hunk 2 above)
Comment (xiaoxmeng) [reply velox/exec/HashBuild.cpp:503]: Removed

File: velox/exec/tests/JoinFuzzerRunner.h (hunk 1 above)
Comment (mbasmanova) [reply velox/exec/tests/JoinFuzzerRunner.h:72]: @xiaoxmeng Would you answer this question?

File: velox/exec/tests/JoinFuzzer.cpp [hunk 6]
```diff
@@ -283,12 +363,25 @@ std::vector<RowVectorPtr> flatten(const std::vector<RowVectorPtr>& vectors) {
...
-    builder.splits(nodeId, nodeSplits);
+  for (const auto& [planNodeId, nodeSplits] : plan.splits) {
+    builder.splits(planNodeId, nodeSplits);
+  }
+
+  if (plan.executionStrategy == core::ExecutionStrategy::kGrouped) {
+    builder.executionStrategy(core::ExecutionStrategy::kGrouped);
+    builder.groupedExecutionLeafNodeIds({plan.probeScanId, plan.buildScanId});
+    builder.numSplitGroups(plan.numGroups);
+    builder.numConcurrentSplitGroups(1);
```
Comment (mbasmanova) [inline velox/exec/tests/JoinFuzzer.cpp:384]: Would it makes sense to test with > 1 concurrent split group as well?

**Review by mbasmanova — APPROVED** [review]
@xiaoxmeng Looks good to me. Thank you for extending the Join Fuzzer to cover grouped execution. Please, run the fuzzer long enough to make sure there are no issues. Appreciate tidying up the code, but, please, next time avoid mixing cleanup / refactoring with material changes. It makes it difficult to review the changes.
